tests/ecmp/inner_hashing/conftest.py

```diff
@@ -0,0 +1,356 @@
+def hash_keys():
+    hash_keys = HASH_KEYS[:]
+    return hash_keys
+
+
+@pytest.fixture(scope="module")
+def symmetric_hashing(duthosts, rand_one_dut_hostname):
+    duthost = duthosts[rand_one_dut_hostname]
+    symmetric_hashing = False
+
+    if duthost.facts['asic_type'] in ["mellanox"]:
+        symmetric_hashing = True
+
+    return symmetric_hashing
+
+
+@pytest.fixture(scope="module", params=IP_VERSIONS_LIST)
+def outer_ipver(request):
+    return request.param
+
+
+@pytest.fixture(scope="module", params=IP_VERSIONS_LIST)
+def inner_ipver(request):
+    return request.param
+
+
+@pytest.fixture(scope="module")
+def config_pbh_table(duthost, vlan_ptf_ports, tbinfo):
+    test_intfs_str = get_dut_test_intfs_str(duthost, vlan_ptf_ports, tbinfo)
+
+    duthost.command(ADD_PBH_TABLE_CMD.format(TABLE_NAME,
+                                             test_intfs_str,
+                                             TABLE_DESCRIPTION))
+
+    yield
+
+    duthost.command(DEL_PBH_TABLE_CMD.format(TABLE_NAME))
+
+
+def get_dut_test_intfs_str(duthost, vlan_ptf_ports, tbinfo):
+    test_intfs = []
+    # get ports according to chosen ptf ports indices
+    mg_facts = duthost.get_extended_minigraph_facts(tbinfo)
+    for intf, index in mg_facts['minigraph_ptf_indices'].items():
+        if index in vlan_ptf_ports:
+            test_intfs.append(intf)
+    return ",".join(test_intfs)
+
+
+@pytest.fixture(scope="module")
+def config_hash_fields(duthost):
+    for hash_field, hash_field_params_dict in HASH_FIELD_CONFIG.items():
+        cmd = get_hash_field_add_cmd(hash_field, hash_field_params_dict)
+        duthost.command(cmd)
+
+    yield
+
+    for hash_field in HASH_FIELD_CONFIG.keys():
+        duthost.command(DEL_PBH_HASH_FIELD_CMD.format(hash_field))
+
+
+def get_hash_field_add_cmd(hash_field_name, hash_field_params_dict):
+    cmd = ADD_PBH_HASH_FIELD_CMD.format(hash_field_name,
+                                        hash_field_params_dict["field"],
+                                        hash_field_params_dict["sequence"])
+    if "mask" in hash_field_params_dict:
+        cmd += " --ip-mask '{}'".format(hash_field_params_dict["mask"])
+    return cmd
+
+
+@pytest.fixture(scope="module")
+def config_hash(duthost):
+    duthost.command(ADD_PBH_HASH_CMD.format(HASH_NAME, PBH_HASH_FIELD_LIST))
+
+    yield
+
+    duthost.command(DEL_PBH_HASH_CMD.format(HASH_NAME))
+
+
+@pytest.fixture(scope="module")
+def config_rules(duthost):
+    for inner_ipver in IP_VERSIONS_LIST:
+        config_ipv4_rules(duthost, inner_ipver)
+        config_ipv6_rules(duthost, inner_ipver)
+
+    yield
+
+    for inner_ipver in IP_VERSIONS_LIST:
+        delete_ipv4_rules(duthost, inner_ipver)
+        delete_ipv6_rules(duthost, inner_ipver)
+
+
+def config_ipv4_rules(duthost, inner_ipver):
+    config_vxlan_rule(duthost, " --ip-protocol {}", V4_ETHER_TYPE, "ipv4", inner_ipver)
+    config_nvgre_rule(duthost, " --ip-protocol {}", V4_ETHER_TYPE, "ipv4", inner_ipver)
```

Comment:
I feel like we maybe lacking test coverage for matching on:
gre_key          = h32 "/" h32   ; GRE key (32 bits) 

So would suggest adding a 3rd random packet format with a gre_key match, that way we can get full test coverage.

Reply:
will be covered in the next PR